types/wlr_drm_lease_v1.c

```diff
@@ -0,0 +1,630 @@
+
+	wl_list_insert(&connector->resources, wl_resource_get_link(wl_resource));
+}
+
+static void lease_device_bind(struct wl_client *wl_client, void *data,
+		uint32_t version, uint32_t id) {
+	struct wlr_drm_lease_device_v1 *device = data;
+
+	struct wl_resource *wl_resource  = wl_resource_create(wl_client,
+		&wp_drm_lease_device_v1_interface, version, id);
+
+	if (!wl_resource) {
+		wl_client_post_no_memory(wl_client);
+		return;
+	}
+
+	wl_resource_set_implementation(wl_resource, &lease_device_impl, device,
+			drm_lease_device_v1_handle_resource_destroy);
+
+	int fd = wlr_drm_backend_get_read_only_fd(device->backend);
+	if (fd < 0) {
+		wlr_log(WLR_ERROR, "Unable to get read only DRM fd for leasing");
+		return;
+	}
+
+	wp_drm_lease_device_v1_send_drm_fd(wl_resource, fd);
+	close(fd);
+
+	wl_list_insert(&device->resources, wl_resource_get_link(wl_resource));
+
+	struct wlr_drm_lease_connector_v1 *connector;
+	wl_list_for_each(connector, &device->connectors, link) {
+		drm_lease_connector_v1_send_to_client(connector, wl_client,
+				wl_resource);
+	}
+}
+
+static void handle_output_destroy(struct wl_listener *listener, void *data) {
+	struct wlr_drm_lease_connector_v1 *conn = wl_container_of(listener, conn,
+			destroy);
+
+	wlr_log(WLR_DEBUG, "Destroying connector %s", conn->output->name);
+	if (conn->active_lease) {
+		wlr_drm_lease_v1_revoke(conn->active_lease);
+	}
+	wlr_drm_lease_manager_withdraw_output(conn->device->manager, conn->output);
+}
+
+bool wlr_drm_lease_manager_offer_output(
+		struct wlr_drm_lease_manager *manager, struct wlr_output *output) {
+	assert(manager && output);
+	assert(wlr_output_is_drm(output));
+
+	wlr_log(WLR_DEBUG, "Offering output %s", output->name);
+
+	struct wlr_drm_lease_device_v1 *device = NULL;
+	wl_list_for_each(device, &manager->devices, link) {
+		if (device->backend == output->backend) {
+			break;
+		}
+	}
+
+	if (!device) {
+		wlr_log(WLR_ERROR, "No wlr_drm_lease_device_v1 associated with the "
+				"offered output");
+		return false;
+	}
+
+	struct wlr_drm_lease_connector_v1 *connector;
+	wl_list_for_each(connector, &device->connectors, link) {
+		if (connector->output == output) {
+			wlr_log(WLR_ERROR, "Output %s has already been offered",
+					output->name);
+			return false;
+		}
+	}
+
+	connector = calloc(1, sizeof(struct wlr_drm_lease_connector_v1));
+	if (!connector) {
+		wlr_log(WLR_ERROR, "Failed to allocatr wlr_drm_lease_connector_v1");
+		return false;
+	}
+
+	connector->output = output;
+	connector->device = device;
+
+	wl_signal_add(&output->events.destroy, &connector->destroy);
+	connector->destroy.notify = handle_output_destroy;
+
+	wl_list_init(&connector->resources);
+	wl_list_insert(&device->connectors, &connector->link);
+
+	struct wl_resource *resource;
+	wl_resource_for_each(resource, &device->resources) {
+		drm_lease_connector_v1_send_to_client(
+				connector, wl_resource_get_client(resource), resource);
```

Comment:
The `drm_lease_connector_v1_send_to_client` function can perform the `wl_resource_get_client` call itself, no need for a separate `wl_client` argument.